src/shader_recompiler/ir/ir_emitter.cpp

```diff
@@ -1553,4 +1558,40 @@ void IREmitter::ImageWrite(const Value& handle, const Value& coords, const Value
     Inst(Opcode::ImageWrite, Flags{info}, handle, coords, color);
 }
 
+// Debug print maps to SPIRV's NonSemantic DebugPrintf instruction
+// Renderdoc will hook in its own implementation of the SPIRV instruction
+// Renderdoc accepts format specifiers, e.g. %u, listed here:
+// https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/main/docs/debug_printf.md
+//
+// Example usage:
+//         ir.DebugPrint("invocation xyz: (%u, %u, %u)",
+//                      {ir.GetVectorReg(IR::VectorReg::V0),
+//                       ir.GetVectorReg(IR::VectorReg::V1),
+//                       ir.GetVectorReg(IR::VectorReg::V2)});
+void IREmitter::DebugPrint(std::string_view format,
+                           boost::container::small_vector<Value, 5> format_args) {
+    constexpr size_t PRINT_MAX_ARGS = NumArgsOf(IR::Opcode::DebugPrint);
+    std::array<Value, PRINT_MAX_ARGS> args;
+
+    ASSERT_MSG(format_args.size() < PRINT_MAX_ARGS, "DebugPrint only supports up to {} format args",
+               PRINT_MAX_ARGS);
+
+    for (int i = 0; i < format_args.size(); i++) {
+        args[i] = format_args[i];
+    }
+
+    for (int i = format_args.size(); i < PRINT_MAX_ARGS; i++) {
+        args[i] = Inst(Opcode::Void);
+    }
+
+    Value val = Inst(Opcode::DebugPrint, args[0], args[1], args[2], args[3], args[4]);
+    DebugPrintFlags flags;
+
+    flags.string_idx.Assign(info.string_pool.size());
+    info.string_pool.emplace_back(format);
```

Comment:
Thinking about it more, if the format string is always constant you don't need a pool and can link to the const char* directly.

Reply:
i changed it to this way so i can fit a 5th argument by stashing a u16 index instead of using arg0 for a pointer

Reply:
i guess it would be simpler though to assume a constant string literal. I wouldn't need to pass Info around as much

Reply:
You could also add a const char* type into the IR::Value tagged union and sacrifice arg0 to pass the format string

Reply:
ok, i prefer that too, even if 4 arg limitation. ill change it